File: 0078-subsets/0078-subsets.rb
# @param {Integer[]} nums
# @return {Integer[][]}
def subsets(nums)
    n = nums.length 

    res = []

    subset_helper(nums, 0, [], res)

    return res 
end 

def subset_helper(nums, i, slate, res)
    if i == nums.length
        res << slate.dup 
        return 
    end 

    subset_helper(nums, i+1, slate, res)

    slate << nums[i]
    subset_helper(nums, i+1, slate, res)
    slate.pop 
end 


 

File: 0078-subsets/0078-subsets.py
class Solution:
    def subset_helper(self, nums, i, slate, res):
        if i == len(nums):
            res.append(slate.copy())
            return 

        self.subset_helper(nums, i+1, slate, res)

        slate.append(nums[i])
        self.subset_helper(nums, i+1, slate, res)
        slate.pop()

        
    def subsets(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)

        res = []

        self.subset_helper(nums, 0, [], res)

        return res 
        